fv/bpf_attach_test.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fv
+
+import (
+	"regexp"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/felix/fv/infrastructure"
+	"github.com/projectcalico/libcalico-go/lib/apiconfig"
+)
+
+var _ = infrastructure.DatastoreDescribe("_BPF-SAFE_ Felix bpf reattach object", []apiconfig.DatastoreType{apiconfig.EtcdV3}, func(getInfra infrastructure.InfraFactory) {
+	var (
+		infra   infrastructure.DatastoreInfra
+		felixes []*infrastructure.Felix
+	)
+
+	BeforeEach(func() {
+		infra = getInfra()
+		//opts := infrastructure.DefaultTopologyOptions()
+		opts := infrastructure.TopologyOptions{
+			FelixLogSeverity: "debug",
+			DelayFelixStart:  true,
+			ExtraEnvVars: map[string]string{
+				"FELIX_BPFENABLED":              "true",
+				"FELIX_DEBUGDISABLELOGDROPPING": "true",
+			},
+		}
+
+		felixes, _ = infrastructure.StartNNodeTopology(1, opts, infra)
+
+		err := infra.AddAllowToDatastore("host-endpoint=='true'")
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
+			infra.DumpErrorData()
+		}
+
+		for _, felix := range felixes {
+			felix.Stop()
+		}
+
+		infra.Stop()
+	})
+
+	It("should not reattach bpf programs", func() {
+		felix := felixes[0]
+
+		// This should not happen at initial execution of felix, since there is no program attached
+		firstRunBase := felix.WatchStdoutFor(regexp.MustCompile("Program already attached, skip reattaching"))
+		// These should happen at first execution of felix, since there is no program attached
+		firstRunProg1 := felix.WatchStdoutFor(regexp.MustCompile("Continue with attaching BPF program to_hep_debug.o"))
+		firstRunProg2 := felix.WatchStdoutFor(regexp.MustCompile("Continue with attaching BPF program from_hep_fib_debug.o"))
+		log.Info("Starting Felix")
+		felix.TriggerDelayedStart()
+		Eventually(firstRunBase, "10s", "100ms").ShouldNot(BeClosed())
```

Comment:
This assertion doesn't look right, it will wait up to 10s for the channel to still be open but it will return early if it ever sees the channel open.
```
Eventually(firstRunBase, "10s", "100ms").ShouldNot(BeClosed())
```
When you're testing a negative like this you should do one of a couple of things:

* Test at a suitable time when some other condition has already been met.  For example:
```
Eventually(firstRunProg1, "10s", "100ms").Should(BeClosed())
Eventually(firstRunProg2, "10s", "100ms").Should(BeClosed())
Expect(firstRunBase).NotTo(BeClosed())  // After the eventually asserts so that it's naturally delayed until felix shoulod have emitted this log. 
```
* Use `Consistently` (returns early only if the assertion fails, i.e. if the channel was unexpectedly closed) rather than `Eventually`

Reply:
Right, thanks for the explanation. Fixed it.